File: guests/ocaml/undo.ml
(* The undo scene, OCaml port — guests/rust/undo.rs, tools/scenes/undo.steps. *)

open Kaya_wire
open Kaya_app

type todo = { title : string } [@@deriving kaya_gen]

(* kaya invents no name for a typing episode (docs/undo-plan.md D8). *)
let what label = if label = "" then "typing" else label

(* A [Map], because [note_list] renders ASCENDING BY KEY and one script is
   compared byte-for-byte across every lane. *)
module Notes = Map.Make (Int64)

let row_key path =
  match path with
  | I64 n :: _ -> n
  | _ -> invalid_arg "kaya: undo scene expects minted (I64) keys"

let put_note notes key text =
  if text = "" then Notes.remove key notes else Notes.add key text notes

let note_list notes =
  match Notes.bindings notes with
  | [] -> "no notes"
  | ns ->
      "notes "
      ^ String.concat ","
          (List.map (fun (key, text) -> Printf.sprintf "%Ld=%s" key text) ns)

(* AN EMPTY PATH IS THE DRAFT; a path names a ROW, whose field has no id an
   app could hold. *)
let fold_texts draft notes texts =
  List.iter
    (fun (t : undo_text) ->
      if t.ut_path = [] then draft := t.ut_text
      else notes := put_note !notes (row_key t.ut_path) t.ut_text)
    texts

let () =
  let app = Kaya_app.create () in

  let draft = ref "" in
  let row_notes = ref Notes.empty in

  build app (fun () ->
      let status = signal (Str "no todos") in
      let history = signal (Str "history empty") in
      let keys = signal (Str "no keys") in
      let notes = signal (Str "no notes") in
      let todos = collection_of todo_record in

      let key_list () =
        let spell (key, _) =
          match key with
          | I64 n -> Int64.to_string n
          | _ -> invalid_arg "kaya: undo scene expects minted (I64) keys"
        in
        match List.map spell (record_items todos) with
        | [] -> "no keys"
        | ks -> "keys " ^ String.concat "," ks
      in

      let field =
        entry ~a11y_id:"draft" ~on_change:(fun text -> draft := text) ()
      in

      let on_add () =
        let d = !draft in
        if d = "" then begin
          let total = count (record_handle todos) in
          write status (Str (Printf.sprintf "nothing to add, %d total" total))
        end
        else begin
          undoable (Printf.sprintf "add %s" d);
          ignore (insert_record_fresh todos { title = d });
          let total = count (record_handle todos) in
          write status (Str (Printf.sprintf "added %s, %d total" d total));
          write keys (Str (key_list ()));
          focus field;
          (* [clear] inside an undoable group is refused at apply
             (docs/undo-plan.md D4). [post], not a nested [build]: a handler
             already is a transaction (tools/check-ambient-tx.py). *)
          post app (fun () -> clear field)
        end
      in

      let on_remove () =
        match record_items todos with
        | [] ->
            let total = count (record_handle todos) in
            write status
              (Str (Printf.sprintf "nothing to remove, %d total" total))
        | (key, todo) :: _ ->
            undoable (Printf.sprintf "remove %s" todo.title);
            remove (record_handle todos) key;
            let total = count (record_handle todos) in
            write status
              (Str (Printf.sprintf "removed %s, %d total" todo.title total));
            write keys (Str (key_list ()))
      in

      let on_star () =
        undoable "star";
        write status (Str "starred")
      in

      let on_focus () = focus field in

      let on_note path text =
        row_notes := put_note !row_notes (row_key path) text;
        write notes (Str (note_list !row_notes))
      in

      let took_back verb step delta =
        (* A programmatic write never echoes, so the delta is the ONLY
           notification for text an undo restored. *)
        fold_texts draft row_notes delta.ud_texts;
        let total = count (record_handle todos) in
        write history
          (Str (Printf.sprintf "%s %s, %d total" verb (what step) total));
        write keys (Str (key_list ()));
        write notes (Str (note_list !row_notes))
      in

      window ~title:"undo"
        ~menus:
          [
            menu ~label:"Edit"
              [
                item ~label:"Undo" ~role:role_undo;
                item ~label:"Redo" ~role:role_redo;
              ];
          ]
        ~on_undone:(took_back "undid") ~on_redone:(took_back "redid") ();

      let root =
        column
          [
            label ~a11y_id:"status" ~bind:status (* label#0 *);
            label ~a11y_id:"history" ~bind:history (* label#1 *);
            label ~a11y_id:"keys" ~bind:keys (* label#2 *);
            label ~a11y_id:"notes" ~bind:notes (* label#3 *);
            w field (* entry#0 *);
            button ~text:"add" ~on_click:on_add (* button#0 *);
            button ~text:"star" ~on_click:on_star (* button#1 *);
            button ~text:"focus" ~on_click:on_focus (* button#2 *);
            button ~text:"remove" ~on_click:on_remove (* button#3 *);
            each (record_handle todos) (fun () ->
                Tpl.(
                  row
                    [
                      label ~bind_field:todo_title;
                      (* UNBOUND ON PURPOSE: a bound field would push the
                         copy's own text back at the widget it came from. *)
                      entry ~on_change:on_note;
                    ]
                    ()));
          ]
          ()
      in
      (* The scene types with real keystrokes, so something must hold focus. *)
      focus field;
      mount root);

  exit (run app)

File: guests/haskell/undo.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DeriveAnyClass #-}

-- The undo scene, Haskell port — guests/rust/undo.rs, tools/scenes/undo.steps.

import Data.Int (Int64)
import Data.IORef (IORef, modifyIORef', newIORef, readIORef, writeIORef)
import qualified Data.Map.Strict as Map
import Data.Maybe (fromMaybe)
import GHC.Generics (Generic)

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

data Todo = Todo {title :: Text}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)


-- | kaya invents no label for a typing episode (docs/undo-plan.md D8).
what :: Text -> Text
what label = if T.null label then "typing" else label

-- | Every key the collection holds, in order.
keyList :: RecordCollection Todo -> Build Text
keyList todos = do
  entries <- recordItems todos
  let ks = map (keyText . fst) entries
  return (if null ks then "no keys" else "keys " <> T.intercalate "," ks)

-- | The app's copy of what is typed in the ROWS: a note per todo key.
type Notes = Map.Map Int64 Text

-- | The row a stamped copy's occurrence names: for a top-level For the
-- path is one key, the todo's own.
rowKey :: [Key] -> Int64
rowKey (k : _) =
  fromMaybe
    (error ("kaya: a todo's key is a minted number, not " ++ show k))
    (keyInt k)
rowKey [] = error "kaya: a stamped copy's path is never empty"

-- | The notes, rendered: every note the app holds, by key.
noteList :: Notes -> Text
noteList notes
  | Map.null notes = "no notes"
  | otherwise = "notes " <> T.intercalate "," [tshow k <> "=" <> v | (k, v) <- Map.toAscList notes]

-- | One note, folded in. An empty note is no note.
noteAt :: Int64 -> Text -> Notes -> Notes
noteAt key text
  | T.null text = Map.delete key
  | otherwise = Map.insert key text

-- | The empty path is the draft, a path names a row; the run is walked whole.
foldTexts :: IORef Text -> IORef Notes -> [UndoText] -> IO ()
foldTexts draftRef notesRef = mapM_ one
  where
    one t
      | null (utPath t) = writeIORef draftRef (utText t)
      | otherwise = modifyIORef' notesRef (noteAt (rowKey (utPath t)) (utText t))

main :: IO ()
main = kayaMain $ \app -> do
  draftRef <- newIORef ""
  notesRef <- newIORef (Map.empty :: Notes)

  (notes, noteNode) <- buildTx app $ do
    status <- signalText "no todos"
    history <- signalText "history empty"
    -- The shared script reads labels BY INDEX, so the declaration order of
    -- these two is contract.
    keys <- signalText "no keys"
    notes <- signalText "no notes"
    todos <- collectionOf @Todo

    -- Restoring never echoes, so the delta is the ONLY notification (D5).
    let walked verb label delta = do
          foldTexts draftRef notesRef (undoTexts delta)
          noted <- noteList <$> readIORef notesRef
          submitTx app $ do
            total <- count (recordHandle todos)
            writeSignal history (verb <> " " <> what label <> ", " <> tshow total <> " total")
            -- ONE transaction with the history label above: the script reads
            -- them in that order.
            list <- keyList todos
            writeSignal keys list
            writeSignal notes noted

    window
      primary
      [ WTitle "undo",
        WMenus
          [ menu
              "Edit"
              []
              [ item "Undo" [IRole roleUndo],
                item "Redo" [IRole roleRedo]
              ]
          ],
        WOnUndone (walked "undid"),
        WOnRedone (walked "redid")
      ]

    -- Built before the buttons that close over it: Build is a PURE state
    -- monad, so nothing can reach back for it later.
    entryField <- entryOn (writeIORef draftRef) [A11yId "draft"]

    let onAdd = do
          draft <- readIORef draftRef
          if T.null draft
            then submitTx app $ do
              total <- count (recordHandle todos)
              writeSignal status ("nothing to add, " <> tshow total <> " total")
            else do
              undoableTx app ("add " <> draft) $ do
                _ <- insertFresh todos (Todo draft)
                total <- count (recordHandle todos)
                writeSignal status ("added " <> draft <> ", " <> tshow total <> " total")
                list <- keyList todos
                writeSignal keys list
                -- A pure effect rides along and is not restored (A2).
                focusWidget entryField
              -- 'clearWidget' inside a group is refused at apply (D4).
              submitTx app (clearWidget entryField)
        -- Two transactions: a label that quotes the model is read BEFORE the
        -- group it names opens.
        onRemove = do
          first <- buildTx app $ do
            entries <- recordItems todos
            case entries of
              [] -> do
                total <- count (recordHandle todos)
                writeSignal status ("nothing to remove, " <> tshow total <> " total")
                return Nothing
              ((key, todo) : _) -> return (Just (key, title todo))
          case first of
            Nothing -> return ()
            Just (key, name) -> undoableTx app ("remove " <> name) $ do
              remove (recordHandle todos) key
              total <- count (recordHandle todos)
              writeSignal status ("removed " <> name <> ", " <> tshow total <> " total")
              list <- keyList todos
              writeSignal keys list
        onStar = undoableTx app "star" (writeSignal status "starred")
        onFocus = submitTx app (focusWidget entryField)

    -- 'forEach' rather than 'each': the node escapes the template, so the
    -- change handler is registered centrally against it (bottom of file).
    (todoRows, noteNode) <- forEach (recordHandle todos) $ do
      note <- entry
      _ <- rowOf [label (field @"title" @Todo), pure note]
      return note

    root <-
      column
        [ labelBound status [A11yId "status"], -- label#0
          labelBound history [A11yId "history"], -- label#1
          labelBound keys [A11yId "keys"], -- label#2
          labelBound notes [A11yId "notes"], -- label#3
          pure entryField, -- entry#0
          buttonOn "add" onAdd, -- button#0
          buttonOn "star" onStar, -- button#1
          buttonOn "focus" onFocus, -- button#2
          buttonOn "remove" onRemove, -- button#3
          pure todoRows
        ]
    -- The scene types real keystrokes, so something must hold focus.
    focusWidget entryField
    mount root
    return (notes, noteNode)

  -- Not an undoable group: an ordinary edit is the platform's step.
  onChange app noteNode $ \path text -> do
    modifyIORef' notesRef (noteAt (rowKey path) text)
    noted <- noteList <$> readIORef notesRef
    submitTx app (writeSignal notes noted)
